File: src/main/java/stdlib/Variables.java
package stdlib;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public final class Variables {

    private static Map<String, Value> variables;
    private static final Stack<Map<String, Value>> stack;

    static {
        stack = new Stack<>();
        variables = new HashMap<>();
        variables.put("PI", new NumberValue(Math.PI));
        variables.put("E",  new NumberValue(Math.E));
    }

    public static void push() {
        stack.push(new HashMap<>(variables));
    }

    public static void pop() {
        variables = stack.pop();
    }

    public static boolean isNotExists(String name){
        return !variables.containsKey(name);
    }

    public static Value get(String name) {
        if (isNotExists(name)){
            return NumberValue.ZERO;
        }

        return variables.get(name);

    }

    public static void set(String name, Value value) {
        variables.put(name, value);
    }

}

File: src/main/java/stdlib/Variables.kt
package stdlib

import java.util.HashMap
import java.util.Stack

object Variables {

    private var variables: MutableMap<String, Value>? = null
    private val stack: Stack<Map<String, Value>> = Stack()

    init {
        variables = HashMap()
        variables!!["PI"] = NumberValue(Math.PI)
        variables!!["E"] = NumberValue(Math.E)
    }

    fun push() {
        stack.push(HashMap(variables!!))
    }

    fun pop() {
        variables = stack.pop() as MutableMap<String, Value>?
    }

    fun isNotExists(name: String): Boolean {
        return !variables!!.containsKey(name)
    }

    operator fun get(name: String): Value? {
        return if (isNotExists(name)) {
            NumberValue.NULL
        } else variables!![name]

    }

    operator fun set(name: String, value: Value) {
        variables!![name] = value
    }

}
